ParProcCo/msmapper_aggregator.py

```diff
@@ -0,0 +1,268 @@
+        self.data_dimensions: int
+        self.nxdata_name: str
+        self.nxdata_path_name: str
+        self.nxentry_name: str
+        self.output_data_files: List[Path]
+        self.renormalisation: bool
+        self.signal_name: str
+        self.total_slices: int
+        self.total_volume: np.ndarray
+        self.use_default_axes: bool = False
+
+    def aggregate(self, total_slices: int, aggregation_output_dir: Path, output_data_files: List[Path]) -> Path:
+        """Overrides AggregatorInterface.aggregate"""
+        self._check_total_slices(total_slices, output_data_files)
+        self._renormalise(output_data_files)
+        aggregated_data_file = self._write_aggregation_file(aggregation_output_dir, output_data_files)
+        return aggregated_data_file
+
+    def _check_total_slices(self, total_slices: int, output_data_files: List[Path]) -> None:
+        if type(total_slices) is int:
+            self.total_slices = total_slices
+        else:
+            raise TypeError(f"total_slices is {type(total_slices)}, should be int\n")
+
+        if len(output_data_files) != self.total_slices:
+            raise ValueError(
+                f"Number of output files {len(output_data_files)} must equal total_slices {self.total_slices}")
+
+    def _renormalise(self, output_data_files: List[Path]) -> None:
+        self.output_data_files = output_data_files
+        self._get_nxdata()
+        self._initialise_arrays()
+        self._accumulate_volumes()
+
+    def _initialise_arrays(self) -> None:
+        self._get_all_axes()
+
+        self.axes_mins = [np.inf] * self.data_dimensions
+        self.axes_maxs = [np.NINF] * self.data_dimensions
+
+        for axis_set in self.all_axes:
+            for j, axis in enumerate(axis_set):
+                self.axes_mins[j] = min([min(axis), self.axes_mins[j]])
+                self.axes_maxs[j] = max([max(axis), self.axes_maxs[j]])
+
+        self.all_slices = []
+        for axes, signal_shape in zip(self.all_axes, self.signal_shapes):
+            axes_lengths = tuple(len(axis) for axis in axes)
+            assert axes_lengths == signal_shape, "axes_lengths must equal volumes_array.shape"
+            slices = []
+            for j, axis in enumerate(axes):
+                start = int(round((axis[0] - self.axes_mins[j]) / self.axes_spacing[j]))
+                stop = axes_lengths[j] + start
+                slices.append(slice(start, stop))
+
+            assert len(slices) == self.data_dimensions, "number of slices must equal self.data_dimensions"
+            self.all_slices.append(slices)
+
+        self.accumulator_axis_lengths = []
+        self.accumulator_axis_ranges = []
+
+        for i in range(self.data_dimensions):
+            length = int(round((self.axes_maxs[i] - self.axes_mins[i]) / self.axes_spacing[i])) + 1
+            self.accumulator_axis_lengths.append(length)
+            ranges = [round(x * self.axes_spacing[i] + self.axes_mins[i], 4)
```

Comment:
Can't use `round` here in the general case. Also simpler to use `np.arange(self.accumulator_axis_lengths[i])` and do arithmetic on that